### PR TITLE
Enable overriding template for unpaid button from theme folder

### PR DESCRIPTION
**Pull request by bumi**

### CODE REVIEW

File: includes/class-bln-publisher-paywall.php [hunk 1]
```diff
@@ -282,13 +282,44 @@ protected function format_paid()
...
+         *
+         * @param $template string  Absolute path to template file
+         * @param $this     object  Plugin instance
+         */
+        $template = apply_filters( 'bln_paywall_unpaid_button_template', $template, $this );
+
+        error_log( print_r( $template, true ) );
+        //error_log( print_r( file_exists($template_override), true ) );
+
+        ob_start();
```
Comment (bumi) [inline includes/class-bln-publisher-paywall.php:312]: is it correct that ob_start() to ob_get_clean() captures the output of the include_once? 

so even though the template file prints something we can still return it as function return value? 

nice!

File: includes/class-bln-publisher-paywall.php [hunk 2]
```diff
@@ -282,13 +282,44 @@ protected function format_paid()
...
+            
+            // Class instance with all plugin data
+            $plugin = $this;
+
+            // Load template file
+            include_once $template;
```
Comment (bumi) [inline includes/class-bln-publisher-paywall.php:318]: could this have some unwanted side-effects? or is this ok to simply include the file here?

**Review by bumi — COMMENTED** [review]
this looks great and super flexible. 

so we have a default but everybody could easily write their own template for the button. 

@cryptoteun what do you think?

**Comment by cryptoteun** [comment]
> this looks great and super flexible.
> 
> so we have a default but everybody could easily write their own template for the button.
> 
> @cryptoteun what do you think?

Sorry, it's a very busy period. I think this will work indeed and gives very much flexibilty, great one @Bobz-zg @bumi

**Comment by bumi** [comment]
@cryptoteun thanks for the feedback.

@Bobz-zg I have the 2 questions open. then let's merge it.

**Comment by Bobz-zg** [comment]
@bumi lmk if any other open questions left here

**Comment by bumi** [comment]
@Bobz-zg did you see my comments: 
https://github.com/getAlby/lightning-publisher-wordpress/pull/27#discussion_r1123001582
https://github.com/getAlby/lightning-publisher-wordpress/pull/27#discussion_r1123002244

**Comment by Bobz-zg** [comment]
Yes, I have replied in a thread below each comment.

**Comment by bumi** [comment]
@Bobz-zg hmm, these ones? 
CleanShot 2023-04-01 at 17 39 47@2x
 I don't see anything.

**Comment by Bobz-zg** [comment]
Not sure do you see my comments here:
Screenshot 2023-04-08 at 08 30 43

**Comment by bumi** [comment]
@Bobz-zg ah, I keep missing that, too. You have to submit the review - see the "pending" badge at your comment. 
This combines multiple comments in one review. ("Files chaged" -> "Review changes" )

File: includes/class-bln-publisher-paywall.php (hunk 2 above)
Comment (bumi) [reply includes/class-bln-publisher-paywall.php:318]: just looks unusual, as don't have much PHP experience these days.

**Comment by Bobz-zg** [comment]
> @Bobz-zg could you add some documentation to the readme and maybe also in the help page within the plugin.
> Explaining where users need to put the template and also providing an example (maybe just a link to an example template)

@bumi How about I add this to WiKi instead? Readme will get bulky very quickly I'm afraid.
We can then update readme to point to Wiki and keep everything organised there.

In general I always look at what [WooCommerce](https://github.com/woocommerce/woocommerce/wiki) is doing, and try to follow their patterns.
Structure would be like:
- Installation
- Shortcodes
- Hooks
- - Actions
- - Filters
- REST API

**Comment by bumi** [comment]
Yeah, we just don't have the wiki so far but we have the readme and I thought it is the easiest step.

**Comment by Bobz-zg** [comment]
@bumi Updated readme and fixed an issue I was not aware of.
I was passing `$plugin_instance` as a 2nd argument in filter, but turns out it's not possible to access private/protected methods from this passed, it's considered like accessing from global scope, so I have removed that argument from filter.

However you can still access it via `$plugin` inside of the template
Weird!

Also, I noticed that Js selector for button is `"[data-lnp-postid] button.wp-lnp-btn"`
 should be wrapped in element with attribute `data-lnp-postid` and needs to be html element `button` with class `wp-lnp-btn`.
I'll add this to readme, it's important if we are going to keep it this way?

**Comment by bumi** [comment]
ah cool. great!

regarding the JS: let's update that we use only specific classes then we don't have the html button requirement and people are free.

**Comment by Bobz-zg** [comment]
Updated, but there is still some required markup, I have added this to readme.
- querySelector is now `.wp-lnp-btn`.
- I have moved post_id to localize_script from the data attribute, `window.LN_Paywall.post_id`

**Comment by Bobz-zg** [comment]
@bumi If there is nothing else here to be done, this is good to go, but you'll need to manually resolve the conflict.

**Comment by bumi** [comment]
@Bobz-zg how do I best resolve the conflict? do we need to do this manually? 
or can we delete the lock file and update it?

**Comment by Bobz-zg** [comment]
@bumi I've manage to resolve it using online editor here on github. Good to go!

**Pull request merged**